module/purpose/KeyboardWalk/src/KeyboardWalk.cpp

```diff
@@ -0,0 +1,367 @@
+        wrefresh(log_window.get());
+    }
+
+    void KeyboardWalk::update_window(const std::shared_ptr<WINDOW>& window,
+                                     const LogColours& colours,
+                                     const std::string& source,
+                                     const std::string& message,
+                                     const bool& print_level) {
+        std::lock_guard<std::mutex> lock(mutex);
+
+        // Print the message source
+        wprintw(window.get(), source.c_str());
+
+        // Print the log level if it is enabled
+        if (print_level) {
+            // Print it in colour if the functionality is available
+            if (colours_enabled) {
+                wattron(window.get(), COLOR_PAIR(short(colours)));
+            }
+            switch (colours) {
+                case LogColours::TRACE_COLOURS: wprintw(window.get(), "TRACE: "); break;
+                case LogColours::DEBUG_COLOURS: wprintw(window.get(), "DEBUG: "); break;
+                case LogColours::INFO_COLOURS: wprintw(window.get(), "INFO: "); break;
+                case LogColours::WARN_COLOURS: wprintw(window.get(), "WARN: "); break;
+                case LogColours::ERROR_COLOURS: waddwstr(window.get(), L"(╯°□°）╯︵ ┻━┻: "); break;
+                case LogColours::FATAL_COLOURS: waddwstr(window.get(), L"(ノಠ益ಠ)ノ彡┻━┻: "); break;
+            }
+            if (colours_enabled) {
+                wattroff(window.get(), COLOR_PAIR(short(colours)));
+            }
+        }
+
+        // Print the log message and refresh the screen
+        wprintw(window.get(), "%s\n", message.c_str());
+        wrefresh(window.get());
+    }
+
+    void KeyboardWalk::forward() {
+        walk_command.x() += DIFF;
+        update_command();
+        print_status();
+        log<NUClear::INFO>("forward");
+    }
+
+    void KeyboardWalk::left() {
+        walk_command.y() += DIFF;
+        update_command();
+        print_status();
+        log<NUClear::INFO>("left");
+    }
+
+    void KeyboardWalk::back() {
+        walk_command.x() -= DIFF;
+        update_command();
+        print_status();
+        log<NUClear::INFO>("back");
+    }
+
+    void KeyboardWalk::right() {
+        walk_command.y() -= DIFF;
+        update_command();
+        print_status();
+        log<NUClear::INFO>("right");
+    }
+
+    void KeyboardWalk::turn_left() {
+        walk_command.z() += ROT_DIFF;
+        update_command();
+        print_status();
+        log<NUClear::INFO>("turn left");
+    }
+
+    void KeyboardWalk::turn_right() {
+        walk_command.z() -= ROT_DIFF;
+        update_command();
+        print_status();
+        log<NUClear::INFO>("turn right");
+    }
+
+    void KeyboardWalk::kick(LimbID::Value l) {
+        switch (l) {
+            case LimbID::LEFT_LEG: emit<Task>(std::make_unique<Kick>(LimbID::LEFT_LEG), 3); break;
+            case LimbID::RIGHT_LEG: emit<Task>(std::make_unique<Kick>(LimbID::RIGHT_LEG), 3); break;
```

Comment:
does this not make the robot fall over when switching because the walk isn't cleanly stopped?